File: android/src/main/java/com/github/quadflask/react/navermap/RNNaverMapView.java
package com.github.quadflask.react.navermap;

import android.graphics.PointF;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;

import com.airbnb.android.react.maps.ViewAttacherGroup;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.LifecycleEventListener;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeArray;
import com.facebook.react.bridge.WritableNativeMap;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.events.RCTEventEmitter;
import com.naver.maps.geometry.LatLng;
import com.naver.maps.geometry.LatLngBounds;
import com.naver.maps.map.*;
import com.naver.maps.map.util.FusedLocationSource;

import java.util.ArrayList;
import java.util.List;

public class RNNaverMapView extends MapView implements OnMapReadyCallback, NaverMap.OnCameraIdleListener, NaverMap.OnMapClickListener, RNNaverMapViewProps {
    private static final int LOCATION_PERMISSION_REQUEST_CODE = 100;
    private ThemedReactContext themedReactContext;
    private FusedLocationSource locationSource;
    private NaverMap naverMap;
    private NaverMapSdk naverMapSdk;
    private ViewAttacherGroup attacherGroup;
    private long lastTouch = 0;
    private final List<RNNaverMapFeature<?>> features = new ArrayList<>();

    public RNNaverMapView(@NonNull ThemedReactContext themedReactContext, ReactApplicationContext appContext, NaverMapOptions naverMapOptions, Bundle instanceStateBundle) {
        super(ReactUtil.getNonBuggyContext(themedReactContext, appContext), naverMapOptions);
        this.themedReactContext = themedReactContext;
        this.locationSource = new FusedLocationSource(appContext.getCurrentActivity(), LOCATION_PERMISSION_REQUEST_CODE);
        super.onCreate(instanceStateBundle);
        naverMapSdk = NaverMapSdk.getInstance(appContext);
//        super.onStart();
        getMapAsync(this);

        // Set up a parent view for triggering visibility in subviews that depend on it.
        // Mainly ReactImageView depends on Fresco which depends on onVisibilityChanged() event
        attacherGroup = new ViewAttacherGroup(this.themedReactContext);
        LayoutParams attacherLayoutParams = new LayoutParams(0, 0);
        attacherLayoutParams.width = 0;
        attacherLayoutParams.height = 0;
        attacherLayoutParams.leftMargin = 99999999;
        attacherLayoutParams.topMargin = 99999999;
        attacherGroup.setLayoutParams(attacherLayoutParams);
        addView(attacherGroup);
    }

    @Override
    public void onMapReady(@NonNull NaverMap naverMap) {
        this.naverMap = naverMap;
        this.naverMap.setLocationSource(locationSource);
        this.naverMap.setOnMapClickListener(this);
        this.naverMap.addOnCameraIdleListener(this);
        this.naverMap.addOnCameraChangeListener((reason, animated) -> {
            if (reason == -1 && System.currentTimeMillis() - lastTouch > 500) { // changed by user
                WritableMap param = Arguments.createMap();
                param.putInt("reason", reason);
                param.putBoolean("animated", animated);
                emitEvent("onTouch", param);
                lastTouch = System.currentTimeMillis();
            }
        });
        naverMapSdk.flushCache(() -> Log.i("NaverMap", "Map Cache Clean"));
        onInitialized();
    }

    @Override
    public void setCenter(LatLng latLng) {
        getMapAsync(e -> {
            CameraUpdate cameraUpdate = CameraUpdate.scrollTo(latLng).animate(CameraAnimation.Easing);
            naverMap.moveCamera(cameraUpdate);
        });
    }

    @Override
    public void setCenter(LatLng latLng, Double zoom, Double tilt, Double bearing) {
        getMapAsync(e -> {
            CameraPosition cam = naverMap.getCameraPosition();
            double zoomValue = zoom == null ? cam.zoom : zoom;
            double tiltValue = tilt == null ? cam.tilt : tilt;
            double bearingValue = bearing == null ? cam.bearing : bearing;

            naverMap.moveCamera(CameraUpdate.toCameraPosition(new CameraPosition(latLng, zoomValue, tiltValue, bearingValue))
                    .animate(CameraAnimation.Easing));
        });
    }

    @Override
    public void zoomTo(LatLngBounds latLngBounds, int paddingInPx) {
        getMapAsync(e -> {
            CameraUpdate cameraUpdate = CameraUpdate.fitBounds(latLngBounds, paddingInPx)
                    .animate(CameraAnimation.Easing);
            naverMap.moveCamera(cameraUpdate);
        });
    }

    @Override
    public void setTilt(int tilt) {
        getMapAsync(e -> {
            final CameraPosition cameraPosition = naverMap.getCameraPosition();
            naverMap.moveCamera(CameraUpdate.toCameraPosition(
                    new CameraPosition(cameraPosition.target, cameraPosition.zoom, tilt, cameraPosition.bearing)));
        });
    }

    @Override
    public void setBearing(int bearing) {
        getMapAsync(e -> {
            final CameraPosition cameraPosition = naverMap.getCameraPosition();
            naverMap.moveCamera(CameraUpdate.toCameraPosition(
                    new CameraPosition(cameraPosition.target, cameraPosition.zoom, cameraPosition.tilt, bearing)));
        });
    }

    @Override
    public void setZoom(float zoom) {
        getMapAsync(e -> {
            naverMap.moveCamera(CameraUpdate.zoomTo(zoom));
        });
    }

    @Override
    public void setMapPadding(int left, int top, int right, int bottom) {
        getMapAsync(e -> {
            naverMap.setContentPadding(left, top, right, bottom);
        });
    }

    @Override
    public void onInitialized() {
        emitEvent("onInitialized", null);
    }

    @Override
    public void showsMyLocationButton(boolean show) {
        getMapAsync(e -> naverMap.getUiSettings().setLocationButtonEnabled(show));
    }

    @Override
    public void setCompassEnabled(boolean show) {
        getMapAsync(e -> naverMap.getUiSettings().setCompassEnabled(show));
    }

    @Override
    public void setScaleBarEnabled(boolean show) {
        getMapAsync(e -> naverMap.getUiSettings().setScaleBarEnabled(show));
    }

    @Override
    public void setZoomControlEnabled(boolean show) {
        getMapAsync(e -> naverMap.getUiSettings().setZoomControlEnabled(show));
    }

    @Override
    public void setLocationTrackingMode(int mode) {
        getMapAsync(e -> naverMap.setLocationTrackingMode(LocationTrackingMode.values()[mode]));
    }

    @Override
    public void setMapType(NaverMap.MapType value) {
        getMapAsync(e -> naverMap.setMapType(value));
    }

    @Override
    public void setMinZoom(float minZoomLevel) {
        getMapAsync(e -> naverMap.setMinZoom(minZoomLevel));
    }

    @Override
    public void setMaxZoom(float maxZoomLevel) {
        getMapAsync(e -> naverMap.setMaxZoom(maxZoomLevel));
    }

    @Override
    public void setBuildingHeight(float height) {
        getMapAsync(e -> naverMap.setBuildingHeight(height));
    }

    @Override
    public void setLayerGroupEnabled(String layerGroup, boolean enable) {
        getMapAsync(e -> naverMap.setLayerGroupEnabled(layerGroup, enable));
    }

    @Override
    public void setNightModeEnabled(boolean enable) {
        getMapAsync(e -> naverMap.setNightModeEnabled(enable));
    }

    @Override
    public void setLogoMargin(int left, int top, int right, int bottom) {
        getMapAsync(e -> naverMap.getUiSettings().setLogoMargin(left, top, right, bottom));
    }

    @Override
    public void setLogoGravity(int gravity) {
        getMapAsync(e -> naverMap.getUiSettings().setLogoGravity(gravity));
    }

    @Override
    public void setScrollGesturesEnabled(boolean enabled) {
        getMapAsync(e -> naverMap.getUiSettings().setScrollGesturesEnabled(enabled));
    }

    @Override
    public void setZoomGesturesEnabled(boolean enabled) {
        getMapAsync(e -> naverMap.getUiSettings().setZoomGesturesEnabled(enabled));
    }

    @Override
    public void setTiltGesturesEnabled(boolean enabled) {
        getMapAsync(e -> naverMap.getUiSettings().setTiltGesturesEnabled(enabled));
    }

    @Override
    public void setRotateGesturesEnabled(boolean enabled) {
        getMapAsync(e -> naverMap.getUiSettings().setRotateGesturesEnabled(enabled));
    }

    @Override
    public void setStopGesturesEnabled(boolean enabled) {
        getMapAsync(e -> naverMap.getUiSettings().setStopGesturesEnabled(enabled));
    }

    @Override
    public void setLiteModeEnabled(boolean enabled) {
        getMapAsync(e -> naverMap.setLiteModeEnabled(enabled));
    }

    @Override
    public void moveCameraFitBound(LatLngBounds bounds, int left, int top, int right, int bottom) {
        getMapAsync(e -> naverMap.moveCamera(CameraUpdate.fitBounds(bounds, left, top, right, bottom).animate(CameraAnimation.Fly, 500)));
    }

    @Override
    public void addFeature(View child, int index) {
        getMapAsync(e -> {
            if (child instanceof RNNaverMapFeature) {
                RNNaverMapFeature<?> annotation = (RNNaverMapFeature<?>) child;
                annotation.addToMap(this);
                features.add(index, annotation);
                int visibility = annotation.getVisibility();
                annotation.setVisibility(View.INVISIBLE);
                ViewGroup annotationParent = (ViewGroup) annotation.getParent();
                if (annotationParent != null) {
                    annotationParent.removeView(annotation);
                }
                // Add to the parent group
                attacherGroup.addView(annotation);
                annotation.setVisibility(visibility);
            }
        });
    }

    @Override
    public void removeFeatureAt(int index) {
        RNNaverMapFeature<?> feature = features.remove(index);
        feature.removeFromMap();
    }

    @Override
    public int getFeatureCount() {
        return features.size();
    }

    @Override
    public View getFeatureAt(int index) {
        return features.get(index);
    }

    @Override
    public void onCameraIdle() {
        CameraPosition cameraPosition = naverMap.getCameraPosition();

        WritableMap param = Arguments.createMap();
        param.putDouble("latitude", cameraPosition.target.latitude);
        param.putDouble("longitude", cameraPosition.target.longitude);
        param.putDouble("zoom", cameraPosition.zoom);
        param.putArray("contentRegion", ReactUtil.toWritableArray(naverMap.getContentRegion()));
        param.putArray("coveringRegion", ReactUtil.toWritableArray(naverMap.getCoveringRegion()));

        emitEvent("onCameraChange", param);
    }

    @Override
    public void onMapClick(@NonNull PointF pointF, @NonNull LatLng latLng) {
        WritableMap param = Arguments.createMap();
        param.putDouble("x", pointF.x);
        param.putDouble("y", pointF.y);
        param.putDouble("latitude", latLng.latitude);
        param.putDouble("longitude", latLng.longitude);

        emitEvent("onMapClick", param);
    }

    @Override
    public void onDestroy() {
        removeAllViews();
        themedReactContext = null;
        locationSource = null;
        naverMap = null;
        attacherGroup = null;
        for (RNNaverMapFeature<?> feature : features)
            feature.removeFromMap();
        features.clear();
        super.onDestroy();
    }

    private void emitEvent(String eventName, WritableMap param) {
        themedReactContext.getJSModule(RCTEventEmitter.class).receiveEvent(getId(), eventName, param);
    }

    public NaverMap getMap() {
        return naverMap;
    }
}

File: android/src/main/java/com/github/quadflask/react/navermap/RNNaverMapView.kt
package com.github.quadflask.react.navermap

import android.graphics.PointF
import android.os.Bundle
import android.util.Log
import android.view.View
import android.view.ViewGroup
import com.airbnb.android.react.maps2.ViewAttacherGroup
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule.RCTDeviceEventEmitter
import com.facebook.react.uimanager.ThemedReactContext
import com.github.quadflask.react.navermap.feature.RNNaverMapFeature
import com.github.quadflask.react.navermap.util.ReactUtil
import com.github.quadflask.react.navermap.util.toWritableArray
import com.naver.maps.geometry.LatLng
import com.naver.maps.geometry.LatLngBounds
import com.naver.maps.map.*
import com.naver.maps.map.NaverMap.*
import com.naver.maps.map.util.FusedLocationSource

class RNNaverMapView(
  private var themedReactContext: ThemedReactContext?,
  appContext: ReactApplicationContext?,
  private var locationSource: FusedLocationSource?,
  naverMapOptions: NaverMapOptions?
) : MapView(ReactUtil.getNonBuggyContext(themedReactContext, appContext!!)!!, naverMapOptions), OnMapReadyCallback, OnCameraIdleListener, OnMapClickListener, RNNaverMapViewProps {
  private var attacherGroup: ViewAttacherGroup?
  private var lastTouch: Long = 0
  private val features: MutableList<RNNaverMapFeature<*>> = mutableListOf()

  init {
    // Set up a parent view for triggering visibility in subviews that depend on it.
    // Mainly ReactImageView depends on Fresco which depends on onVisibilityChanged() event
    attacherGroup = ViewAttacherGroup(themedReactContext).apply {
      layoutParams = LayoutParams(300, 300).apply {
        width = 300
        height = 300
        leftMargin = 99999999
        topMargin = 99999999
      }
    }
    addView(attacherGroup)
    Log.e("RNN", "RNNaverMapView.init")
  }

  override fun onCreate(bundle: Bundle?) {
    super.onCreate(bundle)
    getMapAsync(this)
  }

  override fun onMapReady(map: NaverMap) {
    map.locationSource = locationSource
    map.onMapClickListener = this
    map.addOnCameraIdleListener(this)
    map.addOnCameraChangeListener { reason: Int, animated: Boolean ->
      if (reason == -1 && System.currentTimeMillis() - lastTouch > 500) { // changed by user
        emitEvent("onTouch", Arguments.createMap().apply {
          putInt("reason", reason)
          putBoolean("animated", animated)
        })
        lastTouch = System.currentTimeMillis()
      }
    }
    map.addOnOptionChangeListener {
      emitEvent("onOptionChange", Arguments.createMap().apply {
        putInt("locationTrackingMode", LocationTrackingMode.values().indexOf(map.locationTrackingMode))
      })
    }
    // instanceStateBundle 로 피쳐들이 복구가 안됨. 임시로 map 다시 세팅
    features.forEach {
      if (it.feature.isAdded) {
        it.removeFromMap()
        it.addToMap(map)
      }
    }

    onInitialized()
    Log.e("RNN", "RNNaverMapView.onMapReady")
  }

  override fun setCenter(latLng: LatLng?) {
    getMapAsync { map ->
      latLng?.let {
        val cameraUpdate = CameraUpdate.scrollTo(it).animate(CameraAnimation.Easing)
        map.moveCamera(cameraUpdate)
      }
    }
  }

  override fun setCenter(latLng: LatLng?, zoom: Double?, tilt: Double?, bearing: Double?) {
    getMapAsync { map ->
      val cam = map.cameraPosition
      val target = latLng ?: cam.target
      val zoomValue = zoom ?: cam.zoom
      val tiltValue = tilt ?: cam.tilt
      val bearingValue = bearing ?: cam.bearing
      map.moveCamera(
        CameraUpdate.toCameraPosition(CameraPosition(target, zoomValue, tiltValue, bearingValue))
          .animate(CameraAnimation.Easing)
      )
    }
  }

  override fun zoomTo(latLngBounds: LatLngBounds, paddingInPx: Int) {
    getMapAsync { map ->
      val cameraUpdate = CameraUpdate.fitBounds(latLngBounds, paddingInPx)
        .animate(CameraAnimation.Easing)
      map.moveCamera(cameraUpdate)
    }
  }

  override fun setTilt(tilt: Int) {
    getMapAsync { map ->
      val cameraPosition = map.cameraPosition
      map.moveCamera(
        CameraUpdate.toCameraPosition(
          CameraPosition(cameraPosition.target, cameraPosition.zoom, tilt.toDouble(), cameraPosition.bearing)
        )
      )
    }
  }

  override fun setBearing(bearing: Int) {
    getMapAsync { map ->
      val cameraPosition = map.cameraPosition
      map.moveCamera(
        CameraUpdate.toCameraPosition(
          CameraPosition(cameraPosition.target, cameraPosition.zoom, cameraPosition.tilt, bearing.toDouble())
        )
      )
    }
  }

  override fun setZoom(zoom: Float) {
    getMapAsync { map -> map.moveCamera(CameraUpdate.zoomTo(zoom.toDouble())) }
  }

  override fun setMapPadding(left: Int, top: Int, right: Int, bottom: Int) {
    getMapAsync { map -> map.setContentPadding(left, top, right, bottom) }
  }

  override fun onInitialized() {
    emitEvent("onInitialized", null)
  }

  override fun showsMyLocationButton(show: Boolean) {
    getMapAsync { map -> map.uiSettings.isLocationButtonEnabled = show }
  }

  override fun setCompassEnabled(show: Boolean) {
    getMapAsync { map -> map.uiSettings.isCompassEnabled = show }
  }

  override fun setScaleBarEnabled(show: Boolean) {
    getMapAsync { map -> map.uiSettings.isScaleBarEnabled = show }
  }

  override fun setZoomControlEnabled(show: Boolean) {
    getMapAsync { map -> map.uiSettings.isZoomControlEnabled = show }
  }

  override fun setLocationTrackingMode(mode: Int) {
    getMapAsync { map -> map.locationTrackingMode = LocationTrackingMode.values()[mode] }
  }

  override fun setMapType(value: MapType?) {
    getMapAsync { map -> map.mapType = value!! }
  }

  override fun setMinZoom(minZoomLevel: Float) {
    getMapAsync { map -> map.minZoom = minZoomLevel.toDouble() }
  }

  override fun setMaxZoom(maxZoomLevel: Float) {
    getMapAsync { map -> map.maxZoom = maxZoomLevel.toDouble() }
  }

  override fun setExtent(bounds: LatLngBounds?) {
    getMapAsync { map -> map.extent = bounds }
  }

  override fun setBuildingHeight(height: Float) {
    getMapAsync { map -> map.buildingHeight = height }
  }

  override fun setLayerGroupEnabled(layerGroup: String, enable: Boolean) {
    getMapAsync { map -> map.setLayerGroupEnabled(layerGroup, enable) }
  }

  override fun setNightModeEnabled(enable: Boolean) {
    getMapAsync { map -> map.isNightModeEnabled = enable }
  }

  override fun setLogoMargin(left: Int, top: Int, right: Int, bottom: Int) {
    getMapAsync { map -> map.uiSettings.setLogoMargin(left, top, right, bottom) }
  }

  override fun setLogoGravity(gravity: Int) {
    getMapAsync { map -> map.uiSettings.logoGravity = gravity }
  }

  override fun setScrollGesturesEnabled(enabled: Boolean) {
    getMapAsync { map -> map.uiSettings.isScrollGesturesEnabled = enabled }
  }

  override fun setZoomGesturesEnabled(enabled: Boolean) {
    getMapAsync { map -> map.uiSettings.isZoomGesturesEnabled = enabled }
  }

  override fun setTiltGesturesEnabled(enabled: Boolean) {
    getMapAsync { map -> map.uiSettings.isTiltGesturesEnabled = enabled }
  }

  override fun setRotateGesturesEnabled(enabled: Boolean) {
    getMapAsync { map -> map.uiSettings.isRotateGesturesEnabled = enabled }
  }

  override fun setStopGesturesEnabled(enabled: Boolean) {
    getMapAsync { map -> map.uiSettings.isStopGesturesEnabled = enabled }
  }

  override fun setLiteModeEnabled(enabled: Boolean) {
    getMapAsync { map -> map.isLiteModeEnabled = enabled }
  }

  override fun moveCameraFitBound(bounds: LatLngBounds, left: Int, top: Int, right: Int, bottom: Int) {
    getMapAsync { map -> map.moveCamera(CameraUpdate.fitBounds(bounds, left, top, right, bottom).animate(CameraAnimation.Fly, 500)) }
  }

  override fun addFeature(child: View?, index: Int) {
    Log.e("RNN", "RNNaverMapView.addFeature(child: ${child?.javaClass?.simpleName}, index: $index)")
    getMapAsync { map ->
      if (child is RNNaverMapFeature<*>) {
        val annotation: RNNaverMapFeature<*> = child
        annotation.addToMap(map)
        features.add(index, annotation)
        val visibility = annotation.visibility
        annotation.visibility = INVISIBLE
        val annotationParent = annotation.parent as ViewGroup?
        annotationParent?.removeView(annotation)
        // Add to the parent group
        attacherGroup!!.addView(annotation)
        annotation.visibility = visibility
        Log.e("RNN", "RNNaverMapView.addFeature(child: ${child.javaClass.simpleName}, index: $index) - DONE ($featureCount)")
      } else {
        Log.e("RNN", "addFeature: not feature: $child")
      }
    }
  }

  override fun removeFeatureAt(index: Int) {
    val feature: RNNaverMapFeature<*> = features.removeAt(index)
    feature.removeFromMap()
    Log.e("RNN", "RNNaverMapView.removeFeatureAt(index: $index)")
  }

  override val featureCount: Int
    get() = features.size

  override fun getFeatureAt(index: Int): View {
    return features[index]
  }

  override fun onCameraIdle() {
    getMapAsync { map ->
      emitEvent("onCameraChange", Arguments.createMap().apply {
        val cameraPosition = map.cameraPosition
        putDouble("latitude", cameraPosition.target.latitude)
        putDouble("longitude", cameraPosition.target.longitude)
        putDouble("zoom", cameraPosition.zoom)
        putArray("contentRegion", map.contentRegion.toWritableArray())
        putArray("coveringRegion", map.coveringRegion.toWritableArray())
      })
    }
  }

  override fun onMapClick(pointF: PointF, latLng: LatLng) {
    emitEvent("onMapClick", Arguments.createMap().apply {
      putDouble("x", pointF.x.toDouble())
      putDouble("y", pointF.y.toDouble())
      putDouble("latitude", latLng.latitude)
      putDouble("longitude", latLng.longitude)
    })
  }

  override fun onDestroy() {
    Log.e("RNN", "RNNaverMapView.onDestroy")
    removeAllViews()
    themedReactContext = null
    locationSource = null
    attacherGroup = null
    for (feature in features) feature.removeFromMap()
    features.clear()
    super.onDestroy()
  }

  private fun emitEvent(eventName: String, param: WritableMap?) {
    themedReactContext
      ?.getJSModule(RCTDeviceEventEmitter::class.java)
      ?.emit(eventName, param)
  }
}
